### PR TITLE
Experiment: Use `wp-env` for e2e test environment.

### PR DESCRIPTION
**Pull request by nerrad**
> **Note: This pull is draft because it should not be merged as is. It is just serving as a reference point for the experiment and proof of concept.**

Recently `@woocommerce/e2e-environment` was created with a custom docker environment etc for running End to End tests. As mentioned in [this comment](https://github.com/woocommerce/woocommerce/issues/25465#issuecomment-601800445), initially there was a desire to utilize the [`@wordpress/env`](https://github.com/WordPress/gutenberg/tree/master/packages/env) for a few reasons:

- It is receiving a lot of attention from a number of developers interested in an easy to setup WordPress environment for plugins and themes.
- It is already highly customizable.
- It offloads the responsibility for maintaining a development for WordPress to the WordPress core project itself (which imo is where it belongs).
- Using this will allow us to be consistent with what other projects in the WordPress ecosystem are using.

It doesn't appear that it was given much review so I thought I'd take a maximum one hour to see what I could spin up and get working and this pull is the result:

- I installed `wp-env` as a package with no extra configuration.
- I modified existing e2e tests to point to the new spun up environment url.

And, it works!

Before I did this pull, I spun up the existing environment and ran tests locally.  The results were:

```bash
Test Suites: 3 failed, 8 passed, 11 total
Tests:       9 failed, 2 skipped, 26 passed, 37 total
Snapshots:   0 total
Time:        274.409s
```

With the tests run against `wp-env` the results were:

```bash
Test Suites: 4 failed, 7 passed, 11 total
Tests:       9 failed, 2 skipped, 26 passed, 37 total
Snapshots:   0 total
Time:        274.012s
```

I haven't looked at the tests but it looks like a number of them might be flaky. But for the purposes of this pull, it shows that using `@wordpress/env` is doable imo.

## To run tests against `wp-env` in this pull:

1. Do the normal setup of Woo core (`npm install && npm run build`).
2. `npm run wp-env start` (from woo core root directory)
3. `npm run test:e2e` (from woo core root directory).

## Action steps:

I propose that work is undertaken to switch to using `@wordpress/env` as the environment powering e2e tests. Some additional thoughts as a part of this work:

- please don't bundle all the `WooCommerce` tests with the `@woocommerce/env` package. WooCommerce core should be responsible for it's own e2e tests and they live in WooCommerce core as opposed to being distributed with the package. This keeps the package weight low for plugins pulling it in to power their e2e tests. 
- please offer e2e configuration and utilities as a _separate_ package from the environment. That way projects can pull in the environment separately for quickly spinning up a WooCommerce install for things _other_ than tests if they want. This follows the same model as what WordPress core has established for package division.
- The package _should_ install the latest released WooCommerce core by default but be configurable (passing through to wp-env) by version.
- The package _should_ install the latest released Storefront theme by default (and be configurable). Let's ensure the _default_ tests are run against our flagship theme.
- As much as possible we should simply pass thru custom configurations that are made available via `@wordpress/env` from consuming code. Essentially `@woocommerce/env` would just be a wrapper around `@wordpress/env` pre-populating with a WooCommerce and Storefront install.

### CODE REVIEW

**Comment by nerrad** [comment]
Just noting some work done in the GB repo to configure their travis test runs to use `wp-env`: https://github.com/WordPress/gutenberg/pull/20280

It looks like there are some issues they encountered (and being worked on) that we probably would hit in our attempts to implement `@wordpress/env` as well. It'd be good to collaborate on those fixes if/when necessary.

**Comment by rrennick** [comment]
> It looks like there are some issues they encountered (and being worked on) that we probably would hit in our attempts to implement @wordpress/env as well. It'd be good to collaborate on those fixes if/when necessary.

@nerrad If you are trying to get this working in your own project, you can check out the e2e setup in https://github.com/woocommerce/woocommerce/tree/4.1.0 when the tests were using a docker container managed in the Travis setup. You'll need to port the extra bits over but at least it shows you the extra pieces you'll need for e2e testing.

**Comment by nerrad** [comment]
> @nerrad If you are trying to get this working in your own project, you can check out the e2e setup in https://github.com/woocommerce/woocommerce/tree/4.1.0 when the tests were using a docker container managed in the Travis setup.

At this point I'm not trying to get this working in my own project. I'm presenting the possibility of switching to use `@wordpress/env` for powering the WooCommerce environment instead of the custom docker configuration (which is why I did the pull in the WooCommerce repo to at least demonstrate its feasibility). I did this because from current communication it doesn't appear that this had been tried yet.

**Comment by rrennick** [comment]
> instead of the custom docker configuration

@nerrad Both a vanilla Docker instance or `wp-env` require a custom configuration for Woo e2e testing. The current direction of this PR is to reimplement the original testing setup while swapping `wp-env` for Docker.

File: tests/e2e/env/config/default.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "url": "http://localhost:8084/",
+  "url": "http://localhost:8888/",
```
Comment (noahtallen) [inline tests/e2e/env/config/default.json:2]: Btw, the default test instance is running on `8889`, which has some options more optimized for tests. (in the future, for example, script debug will be set to false there by default)

**Comment by noahtallen** [comment]
> Both a vanilla Docker instance or wp-env require a custom configuration for Woo e2e testing.

If a custom configuration is set for wp-env (i.e. via `.wp-env.json`), then running the WordPress environment behaves the same everywhere (e.g. `wp-env start` do the same thing here and in Gutenberg, or on a WordPress.com project).

**Comment by rrennick** [comment]
@noahtallen @nerrad The [docker configuration](https://github.com/woocommerce/woocommerce/blob/master/tests/e2e/env/docker-compose.yaml) you are looking to replace is included in the `@woocommmerce/e2e-environment` package vs one being configured local to the projects consuming the package.

If you plan to continue the direction of every project having a locally maintained environment it would be easier to base this POC against [4.1.0](https://github.com/woocommerce/woocommerce/tree/4.1.0) which still used a [local docker setup](https://github.com/woocommerce/woocommerce/blob/4.1.0/docker-compose.yaml).

**Comment by rrennick** [comment]
Closing this @nerrad .

**Pull request closed without merging**